hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterMergeTask.java

```diff
@@ -78,8 +84,29 @@ public void run() {
                 executeMergeTasks(nonCoreTasks);
             }
         } finally {
-            lifecycleService.fireLifecycleEvent(joined ? MERGED : MERGE_FAILED);
+            try {
+                resetNonCoreServices();
```

Comment:
What's the reason of resetting non-core services after merge/join is completed, instead of after gathering merge tasks? Won't this have side-effects, keeping former/stale state while joining a new cluster?